### PR TITLE
[somfytahoma] Improvements to avoid cloud login throttling

### PR DESCRIPTION
**Pull request by octa22**
This PR addresses one of the biggest pains of this binding - too many requests error (cloud throtting problem introduced by changes in Tahoma cloud).
Moreover it fixes another problem with local communication when the local gateway IP changes.

### CODE REVIEW

File: bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaBridgeHandler.java [hunk 1]
```diff
@@ -1132,14 +1143,26 @@ public synchronized void refresh(String url, String stateName) {
...
             } else {
                 logger.debug("Cannot call url: {} with params: {}!", getApiFullUrl(url), urlParameters, e);
                 updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
             }
         } catch (TimeoutException e) {
+            errorsCounter++;
             logger.debug("Timeout when calling url: {} with params: {}!", getApiFullUrl(url), urlParameters, e);
-            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
+            if (errorsCounter > MAX_ERRORS) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
```
Comment (lsiepel) [inline bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaBridgeHandler.java:1164]: You could add a custom message to the status detail. Explaining it encountered too much timeouts? So the end user has more information on what is happening without switching to debug mode.

Comment (octa22) [reply bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaBridgeHandler.java]: OK, done

**Review by lsiepel — APPROVED** [review]
LGTM

**Review by jlaur — CHANGES_REQUESTED** [review]
Thanks for the improvements! I have added a few initial comments.

File: bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaBridgeHandler.java [hunk 2]
```diff
@@ -139,6 +141,12 @@ public class SomfyTahomaBridgeHandler extends BaseBridgeHandler {
     // Cloud fallback
     private boolean cloudFallback = false;
 
+    // Communication errors counter
+    private int errorsCounter = 0;
+
+    // Last login timestamp
+    private Instant lastLoginTimestamp = new Date(0).toInstant();
```
[the remark this answers is not in the document]
Comment (octa22) [reply bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaBridgeHandler.java:148]: OK

File: bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaBridgeHandler.java [hunk 3]
```diff
@@ -240,8 +248,8 @@ public synchronized void login() {
             return;
         }
 
-        if (tooManyRequests) {
-            logger.debug("Skipping login due to too many requests");
+        if (tooManyRequests || new Date().toInstant().minusSeconds(LOGIN_LIMIT_TIME).isBefore(lastLoginTimestamp)) {
```
[the remark this answers is not in the document]
Comment (octa22) [reply bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaBridgeHandler.java:251]: OK

File: bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaBridgeHandler.java [hunk 4]
```diff
@@ -277,6 +285,8 @@ public synchronized void login() {
             SomfyTahomaLoginResponse data = gson.fromJson(response.getContentAsString(),
                     SomfyTahomaLoginResponse.class);
 
+            lastLoginTimestamp = new Date().toInstant();
```
[the remark this answers is not in the document]
Comment (octa22) [reply bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaBridgeHandler.java:288]: OK

File: bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaBridgeHandler.java [hunk 5]
```diff
@@ -1132,14 +1143,27 @@ public synchronized void refresh(String url, String stateName) {
                 updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, "Temporarily banned");
                 setTooManyRequests();
             } else if (isEventListenerTimeout(e)) {
+                logger.debug("Event listener timeout occured", e);
```
[the remark this answers is not in the document]
Comment (octa22) [reply bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaBridgeHandler.java:1146]: OK

File: bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaBridgeHandler.java [hunk 6]
```diff
@@ -1132,14 +1143,27 @@ public synchronized void refresh(String url, String stateName) {
...
                 logger.debug("Cannot call url: {} with params: {}!", getApiFullUrl(url), urlParameters, e);
                 updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
             }
         } catch (TimeoutException e) {
+            errorsCounter++;
             logger.debug("Timeout when calling url: {} with params: {}!", getApiFullUrl(url), urlParameters, e);
-            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
+            if (errorsCounter > MAX_ERRORS) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, "Too many timeouts");
+            }
```
[the remark this answers is not in the document]
Comment (octa22) [reply bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaBridgeHandler.java:1166]: Every successfull sent message resets the counter. See InvokeCallToURL method, please.

**Comment by octa22** [comment]
Thanks for the comments. All of them should be fixed now.

**Comment by lolodomo** [comment]
Is it something enough serious that needs a fast merge and a backport to 4.0 ?
@jlaur : can you end your review ? If not and if the answer ot my first question is "yes", I can review myself quickly.

**Comment by jlaur** [comment]
> @jlaur : can you end your review ?

I cannot review it this weekend.

**Review by lolodomo — APPROVED** [review]
LGTM, thank you

**Comment by lolodomo** [comment]
@jlaur, all your comments were considered.

**Comment by lolodomo** [comment]
@octa22 : let us know if you think it should really be backported to 4.0 branch.

**Comment by octa22** [comment]
> @octa22 : let us know if you think it should really be backported to 4.0 branch.

Yeah, I think users will appreciate the more stable and reliable binding. Can we backport it to 4.0 branch, please?

**Pull request merged**